clang/lib/3C/Constraints.cpp

```diff
@@ -438,13 +435,35 @@ bool Constraints::graphBasedSolve() {
     if (!Res) {
       std::set<VarAtom *> Rest;
       Env.doCheckedSolve(true);
-      for (VarAtom *VA : Conflicts) {
-        assert(VA != nullptr);
-        std::string Rsn = "Bad pointer type solution";
-        Geq *ConflictConstraint = createGeq(VA, getWild(), Rsn);
+      for (auto *Conflict : Conflicts) {
+        Atom *ConflictAtom = Conflict->getTargetNode().getData();
+        assert(ConflictAtom != nullptr);
+        ReasonLoc Rsn1 = Conflict->EdgeConstraint->getReason();
+        // Determine a second from the constraints immediately incident to the
+        // conflicting atom. A future improvement should traverse the
+        // constraint graph to find the contradictory constraints to constant
+        // atoms. See correctcomputation/checkedc-clang#680.
+        auto Succs = Conflict->getTargetNode().getEdges();
+        ReasonLoc Rsn2;
+        for (auto *Succ : Succs) {
+          if (auto *SuccGeq = dyn_cast<Geq>(Succ->EdgeConstraint)) {
+            if (Env.getAssignment(ConflictAtom) ==
+                Env.getAssignment(SuccGeq->getLHS()) ||
+                Env.getAssignment(ConflictAtom) ==
+                    Env.getAssignment(SuccGeq->getRHS())) {
+              Rsn2 = Succ->EdgeConstraint->getReason();
+              break;
+            }
+          }
+        }
+        auto Rsn = ReasonLoc("Inferred conflicting types",
+                             PersistentSourceLoc());
+        Geq *ConflictConstraint = createGeq(ConflictAtom, getWild(), Rsn);
+        ConflictConstraint->addReason(Rsn1);
+        ConflictConstraint->addReason(Rsn2);
         addConstraint(ConflictConstraint);
```

Comment:
One more potential problem I thought of after our exchange at https://github.com/correctcomputation/checkedc-clang/issues/298#issuecomment-924243881: If somehow 3C added a conflict constraint with notes and the following code later overwrote the reason:
https://github.com/correctcomputation/checkedc-clang/blob/aa40fb6b233a47da24123410838552885a5472e9/clang/lib/3C/Constraints.cpp#L121-L130
in principle, we should ensure that any notes are cleared when the primary reason is overwritten.  Currently, I don't think this case can arise because all unwritability constraints are added before the start of constraint solving, which generates the conflict constraints.  But if you're trying to build a general mechanism, addressing this potential problem somehow would be helpful for clarity.